ogr/ogrsf_frmts/xodr/ogrxodrlayerlaneborder.cpp

```diff
@@ -0,0 +1,120 @@
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#include "ogr_api.h"
+#include "ogr_geometry.h"
+#include "ogr_xodr.h"
+
+OGRXODRLayerLaneBorder::OGRXODRLayerLaneBorder(
+    const RoadElements &xodrRoadElements, const std::string &proj4Defn)
+    : OGRXODRLayer(xodrRoadElements, proj4Defn)
+{
+    m_poFeatureDefn =
+        std::make_unique<OGRFeatureDefn>(FEATURE_CLASS_NAME.c_str());
+    m_poFeatureDefn->Reference();
+    SetDescription(FEATURE_CLASS_NAME.c_str());
+
+    OGRwkbGeometryType wkbLineStringWithZ = OGR_GT_SetZ(wkbLineString);
+    m_poFeatureDefn->SetGeomType(wkbLineStringWithZ);
+    m_poFeatureDefn->GetGeomFieldDefn(0)->SetSpatialRef(&m_poSRS);
+
+    OGRFieldDefn oFieldID("ID", OFTInteger);
+    m_poFeatureDefn->AddFieldDefn(&oFieldID);
+
+    OGRFieldDefn oFieldRoadID("RoadID", OFTString);
+    m_poFeatureDefn->AddFieldDefn(&oFieldRoadID);
+
+    OGRFieldDefn oFieldType("Type", OFTString);
+    m_poFeatureDefn->AddFieldDefn(&oFieldType);
+
+    OGRFieldDefn oFieldPred("Predecessor", OFTInteger);
+    m_poFeatureDefn->AddFieldDefn(&oFieldPred);
+
+    OGRFieldDefn oFieldSuc("Successor", OFTInteger);
+    m_poFeatureDefn->AddFieldDefn(&oFieldSuc);
+}
+
+int OGRXODRLayerLaneBorder::TestCapability(const char *pszCap)
+{
+    int result = FALSE;
+
+    if (EQUAL(pszCap, OLCZGeometries))
+        result = TRUE;
+
+    return result;
+}
+
+OGRFeature *OGRXODRLayerLaneBorder::GetNextRawFeature()
+{
+    std::unique_ptr<OGRFeature> feature;
+
+    if (m_laneIter != m_roadElements.lanes.end())
+    {
+        feature = std::make_unique<OGRFeature>(m_poFeatureDefn.get());
+
+        odr::Lane lane = *m_laneIter;
+        odr::Line3D laneOuter = *m_laneLinesOuterIter;
+        std::string laneRoadID = *m_laneRoadIDIter;
+
+        OGRLineString lineString;
+        for (const auto &laneVertex : laneOuter)
+        {
+            lineString.addPoint(laneVertex[0], laneVertex[1], laneVertex[2]);
+        }
+        OGRGeometry *geometry = lineString.MakeValid();
```

Comment:
MakeValid() might return nullptr in case of failure, so this should be tested to avoid a nullptr dereference. Furthermore MakeValid() is only available or GEOS-enabled builds. So the choice is either to make GEOS a build requirement for the driver in ogr/ogrsf_frmts/CMakeLists.txt with ``ogr_dependent_driver(xodr OpenDRIVE "GDAL_USE_OPENDRIVE;GDAL_USE_GEOS")``, or to only call MakeValid() if ``OGRGeometryFactory::haveGEOS()`` and use the original geometry then.
Is it common/expected that XODR datasets return non-valid geometries?

Reply:
It should be rather uncommon that XODR datasets return non-valid geometries. But as we rely on libOpenDRIVE's internal linear approximation of parametric geometries, we should at least validate our derived Simple Feature version. I will make this more robust with your suggestions in mind. I see it reasonable to make GEOS dependency optional.

Reply:
> or to only call MakeValid() if `OGRGeometryFactory::haveGEOS()`

Does there exist any similarly convenient function to test if SFCGAL is available?

Reply:
> I will make this more robust with your suggestions in mind

I omit using `MakeValid()` now because handling of a potentially modified return geometry is not trivial. Instead, just a testing for `IsValid()` is performed. If that fails, the feature's geometry will stay unset and a warning is issued. I am not sure if allowing geometry-less features in such boundary cases is a reasonable approach? Theoretically, such cases should not occur.

Further, GEOS is a required dependency now because we need it for `UnaryUnion()`.

Reply:
> Does there exist any similarly convenient function to test if SFCGAL is available?

I don't think so. But you could had a bool OGRGeometryFactory::haveSFCGAL() if needed next to haveGEOS() in ogr/ogrgeometryfactory.cpp)

> Instead, just a testing for `IsValid()` is performed. If that fails, the feature's geometry will stay unset and a warning is issued

That seems a bit extreme. A lot (most) drivers will report invalid geometries, and let the responsibility to the user to fix them with MakeValid() (ogr2ogr has a -makevalid option for example), or discard them, or just process them as it. So if I were you, I would just report the geometry as obtained from libopendrive, ignoring if it is valid or not